File: app/schemas/supplier.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class InfractionOut(BaseModel):
    model_config = {"from_attributes": True}

    id: UUID
    source: str
    infraction_type: str
    description: str | None
    severity: str
    status: str
    action_taken: str | None
    reported_date: datetime


class SupplierOut(BaseModel):
    model_config = {"from_attributes": True}

    id: UUID
    legal_name: str
    dba: str | None
    duns: str | None
    facility_id: str | None
    health_authority: str | None
    parent_organization: str | None
    registered_entity: str | None
    address: str | None
    status: str
    unified_score: float | None
    score_sources_available: int
    score_sources_total: int
    score_last_updated: datetime | None
    orgbook_match_confidence: float | None
    created_at: datetime
    updated_at: datetime


class SupplierDetail(SupplierOut):
    infractions: list[InfractionOut] = []


class SupplierListOut(BaseModel):
    suppliers: list[SupplierOut]
    total: int
    page: int
    page_size: int


class CsvImportRow(BaseModel):
    legal_name: str = Field(min_length=2, max_length=500)
    dba: str | None = Field(default=None, max_length=500)
    duns: str | None = Field(default=None, max_length=20)
    facility_id: str | None = Field(default=None, max_length=100)


class CsvImportResult(BaseModel):
    imported: int
    failed: int
    errors: list[dict] = []

File: frontend/src/types/supplier.ts
export interface Supplier {
  id: string;
  legal_name: string;
  dba: string | null;
  duns: string | null;
  facility_id: string | null;
  health_authority: string | null;
  parent_organization: string | null;
  registered_entity: string | null;
  address: string | null;
  status: SupplierStatus;
  unified_score: number | null;
  score_sources_available: number;
  score_sources_total: number;
  score_last_updated: string | null;
  orgbook_match_confidence: number | null;
  created_at: string;
  updated_at: string;
}

export type SupplierStatus =
  | "imported"
  | "mapped"
  | "active"
  | "flagged"
  | "monitored"
  | "archived";

export interface Infraction {
  id: string;
  source: string;
  infraction_type: string;
  description: string;
  severity: "critical" | "moderate" | "non_critical";
  status: "pending_review" | "resolved" | "active";
  action_taken: string | null;
  reported_date: string;
  created_at: string;
}

export interface SupplierDetail extends Supplier {
  infractions: Infraction[];
}

export type RiskLevel = "low" | "moderate" | "high" | "unknown";

export interface SupplierListResponse {
  suppliers: Supplier[];
  total: number;
  page: number;
  page_size: number;
}

export function scoreToRiskLevel(score: number | null): RiskLevel {
  if (score === null) return "unknown";
  if (score >= 85) return "low";
  if (score >= 65) return "moderate";
  return "high";
}

export function statusLabel(status: SupplierStatus): string {
  return status.charAt(0).toUpperCase() + status.slice(1);
}

export function riskLevelColor(level: RiskLevel): string {
  switch (level) {
    case "low":
      return "text-secondary";
    case "moderate":
      return "text-tertiary";
    case "high":
      return "text-error";
    case "unknown":
      return "text-outline";
  }
}

export function scoreBarColor(score: number | null): string {
  if (score === null) return "bg-outline";
  if (score >= 85) return "bg-secondary";
  if (score >= 65) return "bg-tertiary";
  return "bg-error";
}
